timemachine/cpp/src/fixed_point.hpp

```diff
@@ -16,3 +16,11 @@ RealType __host__ __device__ __forceinline__ FIXED_TO_FLOAT_DU_DP(unsigned long
 template <typename RealType> RealType __host__ __device__ __forceinline__ FIXED_TO_FLOAT(unsigned long long v) {
     return static_cast<RealType>(static_cast<long long>(v)) / FIXED_EXPONENT;
 }
+
+template <typename RealType> bool inline __device__ __host__ energy_overflowed(RealType u, int &overflow_count) {
+    if (!isfinite(u) || u > static_cast<RealType>(LLONG_MAX / static_cast<long long>(FIXED_EXPONENT))) {
```

Comment:
We need to be very precise here. Let's just make this bulletproof for both overflow and :

1) do `u* FIXED_EXPONENT > LLONG_MAX` since that's what actually happens in the casting code.
2) We need to be precise, because we're comparing u * FIXED_EXPONENT (`Real`) against a LLONG_MAX that's casted into a (`Real`), so we should use `nextafter`

```
RealType test_u = u * FIXED_EXPONENT;
if (isnan(u) || is_positive_infinity(u) || is_overflow(u)) {
   overflow_count++;
   return true;
} elseif (is_negative_infinity(u) || is_underflow(u)) {
   overflow_count--;
   return true;
}
return false;

bool is_overflow(x) {
    return x > nextafter(LLONG_MAX, 0);
}

bool is_underflow(x) {
    return x < nextafter(LLONG_MIN, 0);
}

```

Reply:
Added in https://github.com/proteneer/timemachine/pull/1084/commits/8fd09fcb86325602dd48175ba077a9349a262a6e and https://github.com/proteneer/timemachine/pull/1084/commits/251de0b36f9742781cb5c1771bc55dceb9dfdbaf